### PR TITLE
storage format refactor for preparing verkle trie integration

### PR DESCRIPTION
**Pull request by matkt**
Here's a breakdown of the PR:

- Splitting Bonsai: The PR aims to divide Bonsai into two packages

- Common Classes with a Prefix: The common part will include classes prefixed with "DiffBased." These classes are designed to provide a base for both Bonsai and any future storage format that might use this diff-based approach.

- Bonsai's Specifics: Bonsai will retain its unique features in its own package. This means that while it shares the diff-based infrastructure with the common part, it also has its own specific functionalities that are not shared with other storage format.

- Extension to Verkle: this modification add the possibility of adding "Verkle" as a new storage format based on the diff-based architecture. Like Bonsai, Verkle would use the common diff-based classes but also have its own specific features.

In summary, the pull request proposes restructuring Bonsai to make it more modular and efficient. This not only improves data management within Bonsai but also sets the stage for incorporating new format like Verkle, which can share the common base while having their unique functionalities.

### Thanks for sending a pull request! Have you done the following?

- [x] Checked out our [contribution guidelines](https://github.com/hyperledger/besu/blob/main/CONTRIBUTING.md)?
- [x] Considered documentation and added the `doc-change-required` label to this PR [if updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).
- [x] Considered the changelog and included an [update if required](https://wiki.hyperledger.org/display/BESU/Changelog).
- [x] For database changes (e.g. KeyValueSegmentIdentifier) considered compatibility and performed forwards and backwards compatibility tests

### Most advanced CI tests are deferred until PR approval, but you could:

- [ ] locally run all unit tests via: `./gradlew build`
- [ ] locally run all acceptance tests via: `./gradlew acceptanceTest`
- [ ] locally run all integration tests via: `./gradlew integrationTest`
- [ ] locally run all reference tests via: `./gradlew ethereum:referenceTests:referenceTests`

### CODE REVIEW

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/trie/common/GenesisWorldStateProvider.java [hunk 1]
```diff
@@ -59,8 +59,8 @@ public static MutableWorldState createGenesisWorldState(
    * @return a mutable world state for the Genesis block
    */
   private static MutableWorldState createGenesisBonsaiWorldState() {
-    final CachedMerkleTrieLoader cachedMerkleTrieLoader =
-        new CachedMerkleTrieLoader(new NoOpMetricsSystem());
+    final BonsaiCachedMerkleTrieLoader cachedMerkleTrieLoader =
```
[the remark this answers is not in the document]
Comment (matkt) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/trie/common/GenesisWorldStateProvider.java:62]:
```suggestion
    final BonsaiCachedMerkleTrieLoader bonsaiCachedMerkleTrieLoader =
```

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/trie/forest/pruner/Pruner.java [hunk 2]
```diff
@@ -51,7 +51,7 @@ public class Pruner {
   private final ExecutorService executorService;
 
   @VisibleForTesting
-  Pruner(
+  public Pruner(
```
[the remark this answers is not in the document]
Comment (matkt) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/trie/forest/pruner/Pruner.java:54]:
```suggestion
    protected Pruner(
```

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/trie/diffbased/bonsai/BonsaiAccount.java [hunk 3]
```diff
@@ -14,118 +14,95 @@
...
+    super(
         context,
         address,
         address.addressHash(),
         stateTrieAccount.getNonce(),
         stateTrieAccount.getBalance(),
-        stateTrieAccount.getStorageRoot(),
         stateTrieAccount.getCodeHash(),
-        mutable);
+        !mutable);
```
[the remark this answers is not in the document]
Comment (matkt) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/trie/diffbased/bonsai/BonsaiAccount.java]: modified the code , feel free to check if it's better for you

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/trie/common/GenesisWorldStateProvider.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (matkt) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/trie/common/GenesisWorldStateProvider.java]: done

File: besu/src/main/java/org/hyperledger/besu/controller/BesuControllerBuilder.java [hunk 4]
```diff
@@ -575,10 +575,10 @@ public BesuController build() {
             dataDirectory.toString(),
             numberOfBlocksToCache);
 
-    final CachedMerkleTrieLoader cachedMerkleTrieLoader =
+    final BonsaiCachedMerkleTrieLoader cachedMerkleTrieLoader =
```
[the remark this answers is not in the document]
Comment (matkt) [reply besu/src/main/java/org/hyperledger/besu/controller/BesuControllerBuilder.java]: done

**Comment by matkt** [comment]
> lgtm in general, I've been following this changes pretty closely with the verkle work. Maybe getting some else to look at this refactor would be good too. I like the refactor but I was trying to find a better name for the DiffBase classes. Seems to be the eternal struggle of engineers .

Yes I had a lot of difficulties to find Diffbased ^^ it was the less strange name I found

**Review by gfukushima — APPROVED** [review]
lgtm, DiffBase can be renamed in a follow up PR if and when we find a better name for it.

**Pull request merged**